### PR TITLE
Improve webpack 2 definitions

### PR DESCRIPTION
**Pull request by Jessidhia**
- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

There were some preexisting lint failures, but I didn't try fixing them here.

If changing an existing definition:
- [x] Provide a URL to  documentation or source code which provides context for the suggested changes:
* https://github.com/webpack/webpack/blob/master/lib/webpack.js
* https://github.com/webpack/webpack/blob/master/lib/NamedModulesPlugin.js
* https://github.com/webpack/webpack/blob/master/lib/SourceMapDevToolPlugin.js
* https://github.com/webpack/webpack/blob/master/lib/EvalSourceMapDevToolPlugin.js
* https://github.com/webpack/webpack/blob/master/lib/ModuleFilenameHelpers.js
* https://github.com/webpack/webpack/blob/master/lib/SourceMapDevToolModuleOptionsPlugin.js
* https://github.com/webpack/webpack-sources/blob/master/lib/SourceAndMapMixin.js
* I believe `sourceRoot` is meant to be consumed by the `source-map` module.
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

File: webpack/index.d.ts [hunk 1]
```diff
@@ -506,15 +520,63 @@ declare namespace webpack {
...
 
     interface SourceMapDevToolPluginStatic {
-        new (options: any): Plugin;
+        new (options?: SourceMapDevToolPluginOptions): Plugin;
+    }
+
+    // if string | false | null, maps to the filename option
+    type SourceMapDevToolPluginOptions = string | false | null | {
+        // output filename pattern (false/null to append)
+        filename?: string | false | null,
```
Comment (Kovensky) [inline webpack/index.d.ts]: This is more like "string | falsy", but I don't think people should be giving `0` as the filename.

File: webpack/index.d.ts [hunk 2]
```diff
@@ -506,15 +520,63 @@ declare namespace webpack {
...
+        // source map comment pattern (false to not append)
+        append?: false | string,
```
Comment (Kovensky) [inline webpack/index.d.ts]: `false` is the only valid non-string value.

File: webpack/index.d.ts [hunk 3]
```diff
@@ -506,15 +520,63 @@ declare namespace webpack {
...
+        // fallback used when the moduleFilenameTemplate produces a collision
+        fallbackModuleFilenameTemplate?: string,
+        // test/include/exclude files
+        test?: Condition | Condition[],
+        include?: Condition | Condition[],
+        exclude?: Condition | Condition[]
+        // whether to include the footer comment with source information
+        noSources?: boolean,
+        // the source map sourceRoot ("The URL root from which all sources are relative.")
+        sourceRoot?: string | null,
```
Comment (Kovensky) [inline webpack/index.d.ts]: `string | falsy`

**Comment by sandersn** [comment]
This package doesn't list its owners for some reason. Can previous committers review this change?

@flyingsheep @mohsen1 @use-strict @DmitryEfimenko @rakatyal @tkqubo @tkrotoff

File: webpack/index.d.ts [hunk 4]
```diff
@@ -533,6 +595,10 @@ declare namespace webpack {
         new (paths: RegExp[]): Plugin;
     }
 
+    interface NamedModulesPluginStatic {
+        new (): Plugin;
+    }
```
Comment (sandersn) [inline webpack/index.d.ts]: Possibly because `new NamedModulesPlugin()` will return `NamedModulesPlugin` instead of `Plugin`? What were the symptoms?

**Comment by mohsen1** [comment]
Awesome work @Kovensky!

A few issues I found with our config: 

* `resolve.modules` is missing. it's an array of strings. 
* `resolve.mainFields` is missing. It's an array of strings.
* `webpack.LoaderOptionsPlugin` is missing. 

Look at webpack.js.org website for all possible configurations.

**Comment by paulvanbrenk** [comment]
@Kovensky gentle ping, can you look at the requested changes?

File: webpack/webpack-tests.ts [hunk 5]
```diff
@@ -480,3 +482,7 @@ configuration = {
 		]
 	}
 }
+
+const resolve: webpack.Resolve = {
+    cachePredicate: 'boo' // why does this test _not_ fail!?
```
Comment (Kovensky) [inline webpack/webpack-tests.ts]: Why does this test not fail? The type of `webpack.Resolve` is `webpack.NewResolve | webpack.OldResolve`; only `webpack.NewResolve` has a potential `cachePredicate` key, so by narrowing the RHS should be treated as a declaration of `webpack.NewResolve`, but it's being treated as `any`.

**Comment by vvakame** [comment]
@paulvanbrenk could you review this PR again?
If we can't get response from authors. We should merge this PR.
cc/ @tkqubo

**Comment by mohsen1** [comment]
This looks good.

My PR (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13455) will conflict but that's fine. Changes here are 👍

**Pull request merged**